htlcswitch/switch.go

```diff
@@ -566,37 +559,58 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 	// payment circuit by forwarding the settle msg to the channel from
 	// which htlc add packet was initially received.
 	case *lnwire.UpdateFufillHTLC, *lnwire.UpdateFailHTLC:
-		// Exit if we can't find and remove the active circuit to
-		// continue propagating the fail over.
-		circuit, err := s.circuits.remove(packet.payHash)
-		if err != nil {
-			err := errors.Errorf("unable to remove "+
-				"circuit for payment hash: %v", packet.payHash)
-			log.Error(err)
-			return err
-		}
+		if packet.incomingChanID == (lnwire.ShortChannelID{}) {
+			// Use circuit map to find the link to forward settle/fail to.
+			circuit := s.circuits.LookupByHTLC(packet.outgoingChanID,
+				packet.outgoingHTLCID)
+			if circuit == nil {
+				err := errors.Errorf("Unable to find target channel for HTLC "+
+					"settle/fail: channel ID = %s, HTLC ID = %d",
+					packet.outgoingChanID, packet.outgoingHTLCID)
+				log.Error(err)
+				return err
+			}
 
-		// If this is failure than we need to obfuscate the error.
-		if htlc, ok := htlc.(*lnwire.UpdateFailHTLC); ok && !packet.isObfuscated {
-			htlc.Reason = circuit.ErrorEncrypter.IntermediateEncrypt(
-				htlc.Reason,
-			)
+			// Remove circuit since we are about to complete the HTLC.
+			err := s.circuits.Remove(packet.outgoingChanID,
+				packet.outgoingHTLCID)
+			if err != nil {
+				log.Warnf("Failed to close completed onion circuit: "+
+					"(%s, %d) <-> (%s, %d)",
+					circuit.IncomingChanID, circuit.IncomingHTLCID,
+					circuit.OutgoingChanID, circuit.OutgoingHTLCID)
+			} else {
+				log.Debugf("Closed completed onion circuit: "+
+					"(%s, %d) <-> (%s, %d)",
```

Comment:
The payment hash should also be logged as before. The hash is a global identifier and gives more info that just the chanIDs+htlcIDs.